Improved caching with Docker in CI

mec opened the pull request:
This work is one option to improve the performance of CI here on the Rails template.

There is a long Slack thread here:

https://dxw.slack.com/archives/CD8TQ9NP9/p1731332872232279

This is by no stretch optimised. it is just a quick approach that worked for the DfE Complete team.

We kicked things off with some linter fixes to the Dockerfile and then moved on to deliver the actual GitHub Actions changes, see the commit for details.

With CI running all the same checks in the Github Action we no longer need the `ci` scripts so we remove them.

Things we could still do:

- the name of the image, `test_app` is possibly a variable of some kind
- find a better way to build from the cached layers without going through the whole process again?
- anything else you can see

Commit:
Initial update to CI

We wanted to try and speed up the CI tasks for the Rails Template.

This work introduces the same approach we took to save time on the DfE
Complete project.

We use Docker supplied actions to build the image and cache the layers
to the Github actions cache.

Subsequent jobs are set to use the layers from the cache which, in most
cases, gives us a nice speed boost.

We've split the jobs so they can be run in parallel, as we know the
image is the same for each job they all use the strategy.

We've kept the Shellcheck task separate as this code is not strictly
speaking the application code, so doesn't really need to be inside the
image.

yndajas commented on .github/workflows/continuous-integration.yml:
```diff
@@ -8,18 +8,112 @@ on:
+        name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v3
+      -
+        name: Build and cache image
```
**question:** loading the cache

I think one of the problems with the old caching ([removed here](https://github.com/dxw/rails-template/pull/764)) was that it could never find the cache in the load step ([documented here](https://github.com/dxw/dfsseta-apply-for-landing-ruby/pull/290); [example here](https://github.com/dxw/dfsseta-apply-for-landing-ruby/actions/runs/11603584610/job/32310829126) - see the "Run actions/cache@v3" step). Are we confident that this approach doesn't have a similar issue?

yndajas replied:
Or is the purpose of the caching to have a single build for parallelised steps that follow within a single run (the way you're splitting up the test steps makes me wonder!), rather than to build in the first run and then use the cache on subsequent runs? Or is it both?

mec replied:
Yep, I am confident the cache is hit, explained below.

I think it is both in this context.

I think with `mode=min` we cache the Docker image layers, you can see the layers with SHAs in the cache: https://github.com/dxw/rails-template/actions/caches

The initial build ensure the cache is up-to-date, if there are changes in the early part of the image, we have to build them and it takes longer, but that does not happen often.

Each subsequent build will use the cache, you can see the hit rate for each run in the UI e.g. https://github.com/dxw/rails-template/actions/runs/11795643255/attempts/1 on this one the image had to be cached so the hit rate was 12% but you can see all the subsequent builds are 88%.

I prefer to run things in parallel if I can, it doesn't do much here, but with longer running tasks, it can save a lot of time.

There are plenty of optimisations to make here, every situation is different, but this is the general approach.

yndajas replied:
Nice - thanks for the explainer. Really useful!

Commit:
Basic linting on Dockerfile

Commit:
Remove unused CI scripts

Now that we have all the CI tasks in Github Actions these scripts are
never used.

yndajas commented on .github/workflows/continuous-integration.yml:
```diff
@@ -8,18 +10,129 @@ on:
+    runs-on: ubuntu-latest
+    needs: build-and-cache
+    steps:
+      -
+        name: Checkout
+        uses: actions/checkout@v4
+      -
+        name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v3
+      -
+        name: Build and cache
+        uses: docker/build-push-action@v6
+        with:
+          context: .
+          file: ./Dockerfile
+          build-args: |
+            RAILS_ENV=test
+          push: false
+          load: true
+          tags: app_test:latest
+          cache-from: type=gha
```
**suggestion:** reuse

Speaking of reuse, it might be nice to use some anchors and aliases to DRY up these steps?

Somewhere before `jobs:`?

```yml
post-cache-job-common-steps: &post-cache-job-common-steps
  -
    name: Checkout
    uses: actions/checkout@v4
  -
    name: Set up Docker Buildx
    uses: docker/setup-buildx-action@v3
  -
    name: Build and cache
    uses: docker/build-push-action@v6
    with:
      context: .
      file: ./Dockerfile
      build-args: |
        RAILS_ENV=test
      push: false
      load: true
      tags: app_test:latest
      cache-from: type=gha
```

And in each job:
```yml
    steps:
      <<: *post-cache-job-common-steps
      - [the step(s) unique to this job]
```

yndajas replied:
Although it looks like there are only three of these jobs now - I'm sure there were more earlier!

mec replied:
Yep, see updated PR comment! Feel free to jump in and do those? :)

yndajas replied:
Ooh, I just went to do it and got a warning in my editor: `Property post-cache-job-common-steps is not allowed. yaml-schema: GitHub Workflow`

Looks like YAML anchors aren't yet supported in GitHub actions, but they're coming: https://github.com/actions/runner/issues/1182

mec replied:
Do we think there is another low effort way to reuse the shared steps?

yndajas replied:
I've pushed something that might work, but the actions don't seem to be firing at the moment 🤔

yndajas replied:
Alright, done - what do you think?

mec replied:
Composite actions, very cool! Nice work! 👍

yndajas committed:
DRY up cache loading in CI workflow

This adds a composite action to run the two cache-loading steps that are
reused across three of the jobs in the continuous integration workflow.
The second of these steps is a little long and detailed, and differs
slightly but meaningfully from a similar step in the build cache job, so
it might be useful to DRY this up. This also allows us to see the meat
of the post-cache jobs a little easier in the continuous integration
workflow

The `actions/checkout@v4` step is needed in each job in order to load
our action (and presumably also the external ones used in the composite
action)

It would be quite nice to use a YAML anchor or alias to do this kind of
reuse, but these are currently unsupported in GitHub Actions. They might
be on the way soon, so watch this space: https://github.com/actions/runner/issues/1182

Commit:
Remove independent build and cache job

Our caching strategy is to cached across CI jobs with the focus on
caching the layers in the Docker image.

As each job has to build the image from (using the cache) regardless, we
think we can get rid of the independent build and cache job and let the
caching happen as we run each job.

The only way to see if this pays off is to merge the changes and run
another CI workflow to see how much caching we get.

mec commented:
I've dropped the independent 'build and cache' job, which gives us big time savings, I feel like there was a reason I separated that job out, but I think we'll need to merge this to test the caching on another PR.

jdudley1123 commented:
@mec I'd like to get these improvements merged before you leave. Is that doable?

mec commented:
@jdudley1123 I would think so, lets give it a try!

mec commented:
@jdudley1123 if we merge this, we'll want to swing back and update the branch protection rules.

jdudley1123 commented on .github/actions/continuous-integration/load-cache/action.yml:
```diff
@@ -0,0 +1,20 @@
+name: Load cache
+description: Loads the cached image in a post-cache continuous integration step
```
I don't understand what "post-cache continuous integration step" means. Isn't this happening before the step?

mec replied:
Yeah, perhaps changing the name to 'Build and cache image' is enough.

yndajas replied:
Does this change mean that it'll recache after every step? My understanding was that [these lines](https://github.com/dxw/rails-template/pull/803/commits/56915a20d15b486c5e229e0c6861e247fe201e29#diff-c471496bcb3ed47397b73d2e3e3aa41a8a679c86accaed9e05676fd9d5987434R23-R33) built the cache, then this shared action was just loading it. In the [original non-DRY version](https://github.com/dxw/rails-template/pull/803/commits/5f83437b622126c1e2552a19d512fc19a1ee00b8), the last few lines of the first cache step were different to the last few lines of subsequent cache steps

Now it looks like it's a single build/read step/action for each CI step. Is it smart enough to work out whether it needs to build or load, meaning we don't need a different definition for the first and subsequent cache steps?

jdudley1123 commented on .github/workflows/continuous-integration.yml:
```diff
@@ -10,57 +10,42 @@ on:
         uses: ./.github/actions/continuous-integration/load-cache
       -
-        name: Run linters and formatters
+        name: Run Ruby standard
```
I think these changes to the linters should be a separate commit

jdudley1123 reviewed (approved):
This looks really good. Every time I left a comment you addressed it in the next commit 😅

Commit:
Split linting and formatting runs

Split each linter and formatter into their own `run` steps, this is for
clarity.
